ansible_collection/hpe/nimble/plugins/module_utils/hpe_nimble.py

```diff
@@ -0,0 +1,283 @@
+
+
+# remove unchanged item from kwargs by matching them with the data present in given object attrs
+def remove_unchanged_or_null_args(obj_attrs, **kwargs):
+    params = remove_null_args(**kwargs)
+    tosearch = params.copy()
+    changed_attrs_dict = {}
+    for key, value in tosearch.items():
+        server_value = obj_attrs.attrs.get(key)
+
+        if type(server_value) is list and type(value) is dict:
+            # we will land here if the user wants to update a metadata.
+            # server return a list of metadata dictionary
+            temp_server_metadata_dict = {}
+            for server_entry in server_value:
+                temp_server_metadata_dict[server_entry['key']] = server_entry['value']
+            if is_dict_same(temp_server_metadata_dict, value) is False:
+                changed_attrs_dict[key] = value
+                continue
+
+        elif type(server_value) is dict and type(value) is dict:
+            if is_dict_same(server_value, value) is False:
+                changed_attrs_dict[key] = value
+                continue
+
+        elif type(server_value) is list and type(value) is list:
+            # check if the list has dictionary to compare
+            for entry_to_check in value:
+                if type(entry_to_check) is dict:
+                    if is_dict_item_present_on_server(server_value, entry_to_check) is True:
+                        continue
+                    else:
+                        changed_attrs_dict[key] = value
+                        # no need to further check for other keys as we already got one mismatch
+                        break
+                else:
+                    server_value.sort()
+                    value.sort()
+                    if server_value != value:
+                        changed_attrs_dict[key] = value
+                        break
+
+        elif server_value != value:
+            # force is a special key used to force any operation for object
+            # so that is never updated as an attribute
+            if key != "force":
+                changed_attrs_dict[key] = value
+        else:
+            # remove this from param from dictionary as value is same and already present on server
+            params.pop(key)
+    return (changed_attrs_dict, params)
+
+
+def basic_auth_arg_fields():
+
+    fields = {
+        "hostname": {
+            "required": True,
+            "type": "str"
+        },
+        "username": {
+            "required": True,
+            "type": "str",
+            "no_log": True
+        },
+        "password": {
+            "required": True,
+            "type": "str",
+            "no_log": True
+        }
+    }
+    return fields
+
+
+def get_vol_id(client_obj, vol_name):
+    if is_null_or_empty(vol_name):
+        return None
+    else:
+        resp = client_obj.volumes.get(name=vol_name)
+        if resp is None:
+            raise Exception("Invalid value for volume: '%s'" % vol_name)
```

Comment:
Another general comment is using`f-string` for strings formatting. This is the latest standard and provides better performance. Python-SDK follows the same. Please update all the string construction places with f-strings.

`raise Exception(f"Invalid value for volume {vol_name}")`